exonum-java-binding/integration-tests/src/test/java/com/exonum/binding/test/BlockchainIntegrationTest.java

```diff
@@ -0,0 +1,417 @@
+      assertThat(txLocation).isEmpty();
+    });
+  }
+
+  @Test
+  void getBlocks() {
+    testKitTest((blockchain) -> {
+      MapIndex<HashCode, Block> blocks = blockchain.getBlocks();
+      Map<HashCode, Block> blocksMap = toMap(blocks);
+      // Contains both genesis and committed blocks
+      assertThat(blocksMap).hasSize(2);
+      assertThat(blocksMap.get(block.getBlockHash())).isEqualTo(block);
+    });
+  }
+
+  @Test
+  void getBlockByHeight() {
+    testKitTest((blockchain) -> {
+      Block actualBlock = blockchain.getBlock(block.getHeight());
+      assertThat(actualBlock).isEqualTo(block);
+    });
+  }
+
+  @Test
+  void getBlockByInvalidHeight() {
+    testKitTest((blockchain) -> {
+      long invalidBlockHeight = block.getHeight() + 1;
+      Exception e = assertThrows(IndexOutOfBoundsException.class,
+          () -> blockchain.getBlock(invalidBlockHeight));
+      String expectedMessage = String.format("Block height (%s) is out of range [0, %s]",
+          invalidBlockHeight, block.getHeight());
+      assertThat(e).hasMessageContaining(expectedMessage);
+    });
+  }
+
+  @Test
+  void getBlockById() {
+    testKitTest((blockchain) -> {
+      Optional<Block> actualBlock = blockchain.findBlock(block.getBlockHash());
+      assertThat(actualBlock).hasValue(block);
+    });
+  }
+
+  @Test
+  void getLastBlock() {
+    testKitTest((blockchain) -> {
+      Block lastBlock = blockchain.getLastBlock();
+      assertThat(lastBlock).isEqualTo(block);
+    });
+  }
+
+  @Test
+  void getActualConfiguration() {
+    testKitTest((blockchain) -> {
+      StoredConfiguration configuration = blockchain.getActualConfiguration();
+      EmulatedNode emulatedNode = testKit.getEmulatedNode();
+      List<ValidatorKey> validatorKeys = configuration.validatorKeys();
+      short validatorCount = 1;
+      assertThat(validatorKeys).hasSize(validatorCount);
+
+      PublicKey emulatedNodeServiceKey = emulatedNode.getServiceKeyPair().getPublicKey();
+      List<PublicKey> serviceKeys = configuration.validatorKeys().stream()
+          .map(ValidatorKey::serviceKey)
+          .collect(toList());
+
+      assertThat(serviceKeys).hasSize(validatorCount);
+      assertThat(serviceKeys).contains(emulatedNodeServiceKey);
+      HashCode zeroHashCode = HashCode.fromBytes(new byte[DEFAULT_HASH_SIZE_BYTES]);
+      assertThat(configuration.previousCfgHash()).isEqualTo(zeroHashCode);
+    });
+  }
+
+  @Test
+  void getTransactionPool() throws Exception {
+    TestService service = testKit.getService(TestService.SERVICE_ID, TestService.class);
+    TransactionMessage message = constructTestTransactionMessage("Test message", testKit);
+    RawTransaction rawTransaction = toRawTransaction(message);
+    service.getNode().submitTransaction(rawTransaction);
+
+    testKitTest((blockchain) -> {
+      KeySetIndexProxy<HashCode> transactionPool = blockchain.getTransactionPool();
+      assertThat(transactionPool.contains(message.hash())).isTrue();
```

Comment:
We can add `describedAs` or resort to the usual trick with `ImmutableSet.copyOf(transactionPool)`:
```java
assertThat(transactionPool.contains(message.hash()))
          .describedAs("pool=%s", transactionPool)
          .isTrue();
```

Reply:
We can't go with `ImmutableSet.copyOf(transactionPool)` because there's also an afterCommit transaction in transaction pool and we don't really want to construct it here as well :(